src/lib/parser.c

```diff
@@ -840,35 +777,239 @@ static RdbStatus listListpackItem(RdbParser *p, BulkInfo *lpInfo) {
+        if (unlikely(res!=RDB_STATUS_OK)) return NULL;
     } else {
-        /* If handleByLevel == RDB_LEVEL_DATA */
+        static unsigned char dummy;
+        *pEndCh = &dummy;
+        int buflen = 32;
+        res = allocFromCache(p, buflen, RQ_ALLOC_APP_BULK, NULL, &binfo);
+        if (unlikely(res!=RDB_STATUS_OK)) return NULL;
+        binfo->len = ll2string(binfo->ref, buflen, sval);
+    }
+    return binfo;
+}
 
-        ret = ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, p->deepIntegCheck, ziplistItemCallback, p);
+RdbStatus hashZiplistPairs(RdbParser *p, BulkInfo *ziplistBulk) {
+    size_t items = 0;
 
-        if (unlikely(!ret)) {
-            RDB_reportError(p, RDB_ERR_ZIP_LIST_INTEG_CHECK, "elementZiplist(): Ziplist integrity check failed");
+    if (unlikely(0 == ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, 1, counterCallback, &items))) {
+        RDB_reportError(p, RDB_ERR_SSTYPE_INTEG_CHECK,
+                        "hashZiplistPairs(): Ziplist integrity check failed");
+        return RDB_STATUS_ERROR;
+    }
+
+    if (unlikely((items & 1))) {
+        RDB_reportError(p, RDB_ERR_SSTYPE_INTEG_CHECK,
+                        "hashZiplistPairs(): Ziplist integrity check failed. Uneven number of items.");
+        return RDB_STATUS_ERROR;
+    }
+
+    if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
+        registerAppBulkForNextCb(p, ziplistBulk);
+        CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleHashZL, ziplistBulk->ref);
+        return RDB_STATUS_OK;
+    }
+
+    unsigned char *iterZL = ziplistIndex(ziplistBulk->ref, 0);
+    while (iterZL != NULL) {
+        unsigned char *field, *value, tmp1, tmp2, *pend1, *pend2;
+        unsigned int fieldLen, valueLen;
+        long long fieldVal, valueVal;
+        BulkInfo *binfo1, *binfo2;
+
+        ziplistGet(iterZL, &field, &fieldLen, &fieldVal);
+        iterZL = ziplistNext(ziplistBulk->ref, iterZL);
+        ziplistGet(iterZL, &value, &valueLen, &valueVal);
+        iterZL = ziplistNext(ziplistBulk->ref, iterZL);
+
+        if (!(binfo1 = allocBulkForEmbeddedString(p, field, fieldLen, fieldVal, &tmp1, &pend1)))
             return RDB_STATUS_ERROR;
-        }
 
-        if (unlikely(p->errorCode != RDB_OK)) {
+        if (!(binfo2 = allocBulkForEmbeddedString(p, value, valueLen, valueVal, &tmp2, &pend2)))
             return RDB_STATUS_ERROR;
-        }
+
+        registerAppBulkForNextCb(p, binfo1);
+        registerAppBulkForNextCb(p, binfo2);
+        CALL_HANDLERS_CB(p,
+                         *pend1 = tmp1; *pend2 = tmp2, // finalize: restore modified char
+                         RDB_LEVEL_DATA,
+                         rdbData.handleHashElement,
+                         binfo1->ref,
+                         binfo2->ref,
+                         items);
     }
+    return RDB_STATUS_OK;
+}
 
-    return nextParsingElement(p, PE_END_KEY);
+RdbStatus hashListPackPairs(RdbParser *p, BulkInfo *lpBulk) {
```

Comment:
i'm not sure the "pairs" part is needed. it's a hash made of a listpack, of course it's pairs.
i think you took the "pairs" part from redis, where the function wasn't type specific.